internal/event/machineevent/machineevent.go

```diff
@@ -0,0 +1,140 @@
+// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and IronCore contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package machineevent
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/gogo/protobuf/proto"
+	irievent "github.com/ironcore-dev/ironcore/iri/apis/event/v1alpha1"
+	"github.com/ironcore-dev/libvirt-provider/api"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type EventRecorder interface {
+	Eventf(apiMetadata api.Metadata, eventType, reason, message string) error
+	ListEvents() []*irievent.Event
+}
+
+// Options to initialize the machine event store
+type EventStoreOptions struct {
+	MachineEventMaxEvents      int
+	MachineEventTTL            time.Duration
+	MachineEventResyncInterval time.Duration
+}
+
+// EventStore implements the EventRecorder interface and represents an in-memory event store with TTL for events.
+type EventStore struct {
+	maxEvents           int               // Maximum number of events in the store
+	events              []*irievent.Event // Slice of events
+	mutex               sync.Mutex        // Mutex for thread safety
+	eventTTL            time.Duration     // TTL for events
+	eventResyncInterval time.Duration     // Resync interval for event store's TTL expiration check
+	head                int               // Index of the oldest event
+	count               int               // Current number of events in the store
+	log                 logr.Logger       // Logger for logging overridden events
+}
+
+// NewEventStore creates a new EventStore with a fixed number of events and set TTL for events.
+func NewEventStore(log logr.Logger, opts EventStoreOptions) *EventStore {
+	return &EventStore{
+		maxEvents:           opts.MachineEventMaxEvents,
+		events:              make([]*irievent.Event, opts.MachineEventMaxEvents),
+		eventTTL:            opts.MachineEventTTL,
+		eventResyncInterval: opts.MachineEventResyncInterval,
+		head:                0,
+		count:               0,
+		log:                 log,
+	}
+}
+
+// Eventf adds a new Event to the store. Implements the EventRecorder interface.
+func (es *EventStore) Eventf(apiMetadata api.Metadata, eventType, reason, message string) error {
```

Comment:
Can we put the metadata directly in the recording function that we do not need to return an error?

Reply:
Removed the return type for errors, as event recording should not disrupt the main business logic. Instead, errors are now logged directly without being returned.